File: android/src/main/java/com/aspect/webviewbridge/protocol/BridgeProtocol.kt
/**
 * WebView Bridge SDK - 协议层
 *
 * 定义 Bridge 通信协议的基础数据结构
 */

package com.aspect.webviewbridge.protocol

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.JsonNull
import com.google.gson.JsonObject
import com.google.gson.annotations.SerializedName

/**
 * Bridge 请求
 */
data class BridgeRequest(
    @SerializedName("version")
    val version: String,
    
    @SerializedName("type")
    val type: String,
    
    @SerializedName("params")
    val params: JsonObject?,
    
    @SerializedName("callbackId")
    val callbackId: String?
) {
    /**
     * 获取模块名称
     */
    val moduleName: String
        get() = type.substringBefore(".")
    
    /**
     * 获取方法名称
     */
    val methodName: String
        get() = type.substringAfter(".")
    
    /**
     * 获取字符串参数
     */
    fun getString(key: String): String? {
        return params?.get(key)?.takeIf { !it.isJsonNull }?.asString
    }
    
    /**
     * 获取整数参数
     */
    fun getInt(key: String): Int? {
        return params?.get(key)?.takeIf { !it.isJsonNull }?.asInt
    }
    
    /**
     * 获取布尔参数
     */
    fun getBool(key: String): Boolean? {
        return params?.get(key)?.takeIf { !it.isJsonNull }?.asBoolean
    }
    
    /**
     * 获取浮点数参数
     */
    fun getDouble(key: String): Double? {
        return params?.get(key)?.takeIf { !it.isJsonNull }?.asDouble
    }
    
    /**
     * 获取数组参数
     */
    fun getArray(key: String): List<JsonElement>? {
        return params?.get(key)?.takeIf { it.isJsonArray }?.asJsonArray?.toList()
    }
    
    /**
     * 获取字符串数组参数
     */
    fun getStringArray(key: String): List<String>? {
        return getArray(key)?.mapNotNull { 
            if (!it.isJsonNull) it.asString else null 
        }
    }
    
    /**
     * 获取对象参数
     */
    fun getObject(key: String): JsonObject? {
        return params?.get(key)?.takeIf { it.isJsonObject }?.asJsonObject
    }
}

/**
 * Bridge 响应
 */
data class BridgeResponse(
    @SerializedName("callbackId")
    val callbackId: String?,
    
    @SerializedName("code")
    val code: Int,
    
    @SerializedName("msg")
    val msg: String,
    
    @SerializedName("data")
    val data: Any?
) {
    companion object {
        private val gson = Gson()
        
        /**
         * 创建成功响应
         */
        fun success(callbackId: String?, data: Any? = null): BridgeResponse {
            return BridgeResponse(
                callbackId = callbackId,
                code = BridgeErrorCode.SUCCESS.code,
                msg = "success",
                data = data
            )
        }
        
        /**
         * 创建错误响应
         */
        fun error(callbackId: String?, error: BridgeError): BridgeResponse {
            return BridgeResponse(
                callbackId = callbackId,
                code = error.code,
                msg = error.message,
                data = null
            )
        }
    }
    
    /**
     * 转换为 JSON 字符串
     */
    fun toJson(): String {
        return gson.toJson(this)
    }
}

/**
 * Bridge 事件
 */
data class BridgeEvent(
    @SerializedName("eventName")
    val eventName: String,
    
    @SerializedName("data")
    val data: Any?
) {
    companion object {
        private val gson = Gson()
    }
    
    /**
     * 转换为 JSON 字符串
     */
    fun toJson(): String {
        return gson.toJson(this)
    }
}

/**
 * 错误码枚举
 */
enum class BridgeErrorCode(val code: Int, val message: String) {
    // 成功
    SUCCESS(0, "success"),
    
    // 协议错误 (1xxx)
    PARSE_ERROR(1001, "请求解析失败"),
    INVALID_VERSION(1002, "协议版本不支持"),
    INVALID_TYPE(1003, "无效的请求类型"),
    INVALID_PARAMS(1004, "参数错误"),
    TIMEOUT(1005, "请求超时"),
    
    // 能力错误 (2xxx)
    MODULE_NOT_FOUND(2001, "模块不存在"),
    METHOD_NOT_FOUND(2002, "方法不存在"),
    CAPABILITY_NOT_SUPPORTED(2003, "设备不支持此能力"),
    
    // 权限错误 (3xxx)
    PERMISSION_DENIED(3001, "权限被拒绝"),
    PERMISSION_NOT_GRANTED(3002, "未授予权限"),
    PERMISSION_PERMANENTLY_DENIED(3003, "权限被永久拒绝"),
    
    // 设备限制 (4xxx)
    DEVICE_NOT_SUPPORTED(4001, "设备不支持"),
    RESOURCE_UNAVAILABLE(4002, "资源不可用"),
    
    // 内部错误 (5xxx)
    INTERNAL_ERROR(5001, "内部错误"),
    UNKNOWN_ERROR(5999, "未知错误");
}

/**
 * Bridge 错误
 */
data class BridgeError(
    val code: Int,
    override val message: String
) : Exception(message) {
    
    constructor(errorCode: BridgeErrorCode) : this(errorCode.code, errorCode.message)
    
    constructor(errorCode: BridgeErrorCode, customMessage: String) : this(errorCode.code, customMessage)
    
    companion object {
        fun parseError(detail: String? = null): BridgeError {
            return BridgeError(
                BridgeErrorCode.PARSE_ERROR.code,
                detail ?: BridgeErrorCode.PARSE_ERROR.message
            )
        }
        
        fun invalidParams(param: String): BridgeError {
            return BridgeError(
                BridgeErrorCode.INVALID_PARAMS.code,
                "参数错误: $param"
            )
        }
        
        fun moduleNotFound(moduleName: String): BridgeError {
            return BridgeError(
                BridgeErrorCode.MODULE_NOT_FOUND.code,
                "模块不存在: $moduleName"
            )
        }
        
        fun methodNotFound(method: String): BridgeError {
            return BridgeError(
                BridgeErrorCode.METHOD_NOT_FOUND.code,
                "方法不存在: $method"
            )
        }
        
        fun permissionDenied(permission: String): BridgeError {
            return BridgeError(
                BridgeErrorCode.PERMISSION_DENIED.code,
                "权限被拒绝: $permission"
            )
        }
        
        fun internalError(detail: String? = null): BridgeError {
            return BridgeError(
                BridgeErrorCode.INTERNAL_ERROR.code,
                detail ?: BridgeErrorCode.INTERNAL_ERROR.message
            )
        }
        
        fun notSupported(feature: String): BridgeError {
            return BridgeError(
                BridgeErrorCode.CAPABILITY_NOT_SUPPORTED.code,
                "不支持: $feature"
            )
        }
    }
}

File: webview-bridge-sdk/ios/Sources/Protocol/BridgeProtocol.swift
/**
 * WebView Bridge SDK - 协议层
 *
 * 定义 JS ↔ Native 通信协议的核心类型
 * 协议版本: 1.0
 */

import Foundation

// MARK: - 协议版本

/// 当前协议版本
public let BridgeProtocolVersion = "1.0"

// MARK: - 请求结构

/// Bridge 请求结构
public struct BridgeRequest: Codable {
    /// 协议版本
    public let version: String
    /// 能力标识: Module.Method
    public let type: String
    /// 请求参数
    public let params: [String: AnyCodable]
    /// 唯一回调 ID
    public let callbackId: String
    
    /// 解析模块名
    public var moduleName: String {
        let parts = type.split(separator: ".")
        return parts.count >= 1 ? String(parts[0]) : ""
    }
    
    /// 解析方法名
    public var methodName: String {
        let parts = type.split(separator: ".")
        return parts.count >= 2 ? String(parts[1]) : ""
    }
}

// MARK: - 响应结构

/// Bridge 响应结构
public struct BridgeResponse: Codable {
    /// 对应请求的回调 ID
    public let callbackId: String
    /// 错误码（0 表示成功）
    public let code: Int
    /// 人类可读的错误信息
    public let msg: String
    /// 响应数据
    public let data: AnyCodable?
    
    public init(callbackId: String, code: Int, msg: String, data: AnyCodable? = nil) {
        self.callbackId = callbackId
        self.code = code
        self.msg = msg
        self.data = data
    }
    
    /// 创建成功响应
    public static func success(callbackId: String, data: Any? = nil) -> BridgeResponse {
        return BridgeResponse(
            callbackId: callbackId,
            code: BridgeErrorCode.success.rawValue,
            msg: "",
            data: data.map { AnyCodable($0) }
        )
    }
    
    /// 创建错误响应
    public static func error(callbackId: String, code: BridgeErrorCode, msg: String) -> BridgeResponse {
        return BridgeResponse(
            callbackId: callbackId,
            code: code.rawValue,
            msg: msg,
            data: nil
        )
    }
}

// MARK: - 事件结构

/// Native → Web 事件结构
public struct BridgeEvent: Codable {
    /// 事件名称: Module.EventName
    public let event: String
    /// 事件数据
    public let data: AnyCodable?
    
    public init(event: String, data: Any? = nil) {
        self.event = event
        self.data = data.map { AnyCodable($0) }
    }
}

// MARK: - 错误码

/// Bridge 错误码定义
public enum BridgeErrorCode: Int {
    // 成功
    case success = 0
    
    // 1xxx: 协议错误
    case invalidJson = 1001
    case missingRequiredField = 1002
    case invalidVersion = 1003
    case invalidTypeFormat = 1004
    case invalidParams = 1005
    
    // 2xxx: 能力错误
    case moduleNotFound = 2001
    case methodNotFound = 2002
    case capabilityNotSupported = 2003
    
    // 3xxx: 权限错误
    case permissionDenied = 3001
    case permissionNotDetermined = 3002
    case permissionRestricted = 3003
    case permissionRequired = 3004
    
    // 4xxx: 设备/系统限制
    case deviceNotSupported = 4001
    case osVersionTooLow = 4002
    case systemRestriction = 4003
    case featureDisabled = 4004
    
    // 5xxx: 内部异常
    case internalError = 5001
    case timeout = 5002
    case bridgeNotReady = 5003
    case webViewDestroyed = 5004
    case cancelled = 5005
}

// MARK: - Bridge 错误

/// Bridge 错误
public struct BridgeError: Error, LocalizedError {
    public let code: BridgeErrorCode
    public let message: String
    
    public var errorDescription: String? {
        return "[\(code.rawValue)] \(message)"
    }
    
    public init(code: BridgeErrorCode, message: String) {
        self.code = code
        self.message = message
    }
    
    public static func moduleNotFound(_ name: String) -> BridgeError {
        return BridgeError(code: .moduleNotFound, message: "模块不存在: \(name)")
    }
    
    public static func methodNotFound(_ type: String) -> BridgeError {
        return BridgeError(code: .methodNotFound, message: "方法不存在: \(type)")
    }
    
    public static func invalidParams(_ detail: String) -> BridgeError {
        return BridgeError(code: .invalidParams, message: "参数无效: \(detail)")
    }
}

// MARK: - AnyCodable

/// 用于处理任意类型的 Codable 包装器
public struct AnyCodable: Codable {
    public let value: Any
    
    public init(_ value: Any) {
        self.value = value
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if container.decodeNil() {
            self.value = NSNull()
        } else if let bool = try? container.decode(Bool.self) {
            self.value = bool
        } else if let int = try? container.decode(Int.self) {
            self.value = int
        } else if let double = try? container.decode(Double.self) {
            self.value = double
        } else if let string = try? container.decode(String.self) {
            self.value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            self.value = array.map { $0.value }
        } else if let dictionary = try? container.decode([String: AnyCodable].self) {
            self.value = dictionary.mapValues { $0.value }
        } else {
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "无法解码的类型"
            )
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch value {
        case is NSNull:
            try container.encodeNil()
        case let bool as Bool:
            try container.encode(bool)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let array as [Any]:
            try container.encode(array.map { AnyCodable($0) })
        case let dictionary as [String: Any]:
            try container.encode(dictionary.mapValues { AnyCodable($0) })
        default:
            throw EncodingError.invalidValue(
                value,
                EncodingError.Context(codingPath: [], debugDescription: "无法编码的类型: \(type(of: value))")
            )
        }
    }
    
    /// 获取字符串值
    public var stringValue: String? {
        return value as? String
    }
    
    /// 获取整数值
    public var intValue: Int? {
        return value as? Int
    }
    
    /// 获取布尔值
    public var boolValue: Bool? {
        return value as? Bool
    }
    
    /// 获取字典值
    public var dictionaryValue: [String: Any]? {
        return value as? [String: Any]
    }
    
    /// 获取数组值
    public var arrayValue: [Any]? {
        return value as? [Any]
    }
}

// MARK: - 参数提取扩展

public extension Dictionary where Key == String, Value == AnyCodable {
    /// 获取字符串参数
    func getString(_ key: String) -> String? {
        return self[key]?.stringValue
    }
    
    /// 获取整数参数
    func getInt(_ key: String) -> Int? {
        return self[key]?.intValue
    }
    
    /// 获取布尔参数
    func getBool(_ key: String) -> Bool? {
        return self[key]?.boolValue
    }
    
    /// 获取必填字符串参数
    func requireString(_ key: String) throws -> String {
        guard let value = getString(key) else {
            throw BridgeError.invalidParams("缺少必填参数: \(key)")
        }
        return value
    }
}
